File: 74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0)
            return false;
        int left = 0;
        int right = matrix.length * matrix[0].length - 1;
        while (left <= right) {
            int mid = (left + right)/2;
            int row = mid / matrix[0].length;
            int col = mid % matrix[0].length;
            if (matrix[row][col] == target) return true;
            else if (matrix[row][col] > target) right = mid - 1;
            else left = mid + 1;
        }
        return false;
    }
}

// Time: O(log(m*n))
// Space: O(1)

File: amazon/python/74. Search a 2D Matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if matrix is None:
            return False
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        
        m = len(matrix) # nrow
        n = len(matrix[0]) # ncol
        # binary search on all
        l = 0
        r = m * n - 1
        
        while l <= r:
            mid = l + (r - l) // 2 # 5
            mid_r, mid_c = mid//n, mid%n
            cur_val = matrix[mid_r][mid_c]
            if target == cur_val:
                return True
            elif target > cur_val:
                l = mid + 1
            else:
                r = mid - 1
        
        return False
    
    # Time: O(log mn)
    # Space: O(1)
